### PR TITLE
[Serializer][FrameworkBundle] Add a DateInterval normalizer

### PR DESCRIPTION
**Pull request by Lctrs**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/8267

Could be useful for API needing to submit a duration.

Most code have been adapted from @MisatoTremor's DateInterval form type. Credits to him.

### CODE REVIEW

**Comment by Lctrs** [comment]
Failure is unrelated.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 1]
```diff
@@ -1336,12 +1337,19 @@ private function registerSecurityCsrfConfiguration(array $config, ContainerBuild
     private function registerSerializerConfiguration(array $config, ContainerBuilder $container, XmlFileLoader $loader)
     {
         if (class_exists('Symfony\Component\Serializer\Normalizer\DataUriNormalizer')) {
-            // Run after serializer.normalizer.object
+            // Run before serializer.normalizer.object
```
Comment (ogizanagi) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Could you submit another PR to fix this in 3.3 instead?

Comment (Lctrs) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: (y)

Done in #23765.

File: src/Symfony/Component/Serializer/Normalizer/DateIntervalNormalizer.php [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+     * @throws UnexpectedValueException
+     */
+    public function denormalize($data, $class, $format = null, array $context = array())
+    {
+        if (!is_string($data)) {
+            throw new InvalidArgumentException('Data expected to be a string, '.gettype($data).' given.');
+        }
+
+        if (!$this->isISO8601($data)) {
+            throw new UnexpectedValueException('Non ISO 8601 interval strings are not supported yet.');
```
Comment (ogizanagi) [inline src/Symfony/Component/Serializer/Normalizer/DateIntervalNormalizer.php]: Why "yet"? Do you plan to add support for this and how?
On the contrary, I'd change the message for "Expected a valid ISO 8601 interval string." and remove related tests/dataprovider.

**Review by ogizanagi — APPROVED** [review]

File: src/Symfony/Component/Serializer/Normalizer/DateIntervalNormalizer.php (hunk 2 above)
Comment (Lctrs) [inline src/Symfony/Component/Serializer/Normalizer/DateIntervalNormalizer.php]: Well, as I said, most code have been adapted from the DateInterval form type, and especially from the class [DateIntervalToStringTransformer](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/Extension/Core/DataTransformer/DateIntervalToStringTransformer.php) which have the same exception message.

I thought that if non ISO 8601 interval string support was added to this class, maybe we could port it to the DateIntervalNormalizer. But I'm not sure it's worth it or a valid use case for the Serializer part.

Comment (chalasr) [inline src/Symfony/Component/Serializer/Normalizer/DateIntervalNormalizer.php]: Agree with @ogizanagi, let's not promise something that we're not sure to want.

Comment (Lctrs) [inline src/Symfony/Component/Serializer/Normalizer/DateIntervalNormalizer.php]: I agree. 

I'll try to make the requested changes before I go on holidays.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 3]
```diff
@@ -1377,6 +1378,13 @@ private function registerSerializerConfiguration(array $config, ContainerBuilder
             $definition->addTag('serializer.normalizer', array('priority' => -920));
         }
 
+        if (class_exists(DateIntervalNormalizer::class)) {
+            // Run before serializer.normalizer.object
+            $definition = $container->register('serializer.normalizer.dateinterval', DateIntervalNormalizer::class);
+            $definition->setPublic(false);
+            $definition->addTag('serializer.normalizer', array('priority' => -915));
```
Comment (Lctrs) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: @ogizanagi are you okay with this priority ? Or should I change it to -930 to be more consistent with other normalizers ?

**Comment by Lctrs** [comment]
PR updated to fix comment made by @ogizanagi and @chalasr.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php (hunk 3 above)
Comment (ogizanagi) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Looks good to me :)

**Comment by Lctrs** [comment]
Unless there are more comments that need to be addressed, it should be ready to merge then :)

File: src/Symfony/Component/Serializer/Normalizer/DateIntervalNormalizer.php [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    /**
+     * {@inheritdoc}
+     *
+     * @throws InvalidArgumentException
+     * @throws UnexpectedValueException
+     */
+    public function denormalize($data, $class, $format = null, array $context = array())
+    {
+        if (!is_string($data)) {
+            throw new InvalidArgumentException('Data expected to be a string, '.gettype($data).' given.');
```
Comment (ogizanagi) [inline src/Symfony/Component/Serializer/Normalizer/DateIntervalNormalizer.php]: You should use `sprintf` for exception messages.

**Comment by Lctrs** [comment]
Failure doesn't seem to be related.

**Comment by ogizanagi** [comment]
Could you add an entry in component's CHANGELOG.md file?

**Comment by Lctrs** [comment]
@ogizanagi Done.

**Comment by Lctrs** [comment]
Failure once again not related.

**Comment by Lctrs** [comment]
@ogizanagi Is there anything else that prevent it from being merged ?

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 5]
```diff
@@ -631,9 +632,9 @@ private function registerWorkflowConfiguration(array $config, ContainerBuilder $
 
             $transitions = array();
             foreach ($workflow['transitions'] as $transition) {
-                if ($type === 'workflow') {
+                if ('workflow' === $type) {
```
Comment (ogizanagi) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Must be reverted. This change does not belong to this PR.

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php [hunk 6]
```diff
@@ -631,9 +632,9 @@ private function registerWorkflowConfiguration(array $config, ContainerBuilder $
...
                     $transitions[] = new Definition(Workflow\Transition::class, array($transition['name'], $transition['from'], $transition['to']));
-                } elseif ($type === 'state_machine') {
+                } elseif ('state_machine' === $type) {
```
Comment (ogizanagi) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php]: Same

File: src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php [hunk 7]
```diff
@@ -1088,6 +1104,7 @@ private function assertCachePoolServiceDefinitionIsCreated(ContainerBuilder $con
                 if (ChainAdapter::class === $parentDefinition->getClass()) {
                     break;
                 }
+                // no break
```
Comment (ogizanagi) [inline src/Symfony/Bundle/FrameworkBundle/Tests/DependencyInjection/FrameworkExtensionTest.php]: Same

**Comment by ogizanagi** [comment]
@Lctrs : Nothing to me (appart from the comments I've just added, since the PR was tweaked since my last review). We just miss another approval.

ping @symfony/deciders , @dunglas :)

**Comment by ogizanagi** [comment]
Thank you @Lctrs.
